bothan-api/server-cli/src/commands/query.rs

```diff
@@ -0,0 +1,424 @@
+
+    display_asset_infos(asset_infos);
+    Ok(())
+}
+
+async fn query_kraken<T: Into<Duration>>(
+    opts: bothan_kraken::WorkerOpts,
+    query_ids: &[String],
+    timeout: T,
+) -> anyhow::Result<()> {
+    let connector = bothan_kraken::api::WebSocketConnector::new(opts.url);
+    let asset_infos = query_single_websocket(connector, query_ids, timeout.into()).await?;
+
+    display_asset_infos(asset_infos);
+    Ok(())
+}
+
+async fn query_okx<T: Into<Duration>>(
+    opts: bothan_okx::WorkerOpts,
+    query_ids: &[String],
+    timeout: T,
+) -> anyhow::Result<()> {
+    let connector = bothan_okx::api::WebSocketConnector::new(opts.url);
+    let asset_infos = query_single_websocket(connector, query_ids, timeout.into()).await?;
+
+    display_asset_infos(asset_infos);
+    Ok(())
+}
+
+async fn query_rest_api<P: RestAssetInfoProvider>(
+    provider: &P,
+    query_ids: &[String],
+    timeout_interval: Duration,
+) -> anyhow::Result<Vec<AssetInfo>>
+where
+    P::Error: StdError + Send + Sync + 'static,
+{
+    let deduped_ids = dedup_ids(query_ids);
+    Ok(timeout(timeout_interval, provider.get_asset_info(&deduped_ids)).await??)
+}
+
+async fn query_single_websocket<P, E1, E2, C>(
+    connector: C,
+    ids: &[String],
+    timeout: Duration,
+) -> anyhow::Result<Vec<AssetInfo>>
+where
+    E1: StdError + Send + Sync + 'static,
+    E2: StdError + Send + Sync + 'static,
+    P: WebSocketAssetInfoProvider<SubscriptionError = E1, PollingError = E2>,
+    C: AssetInfoProviderConnector<Provider = P, Error = E1>,
+{
+    let mut provider = connector.connect().await?;
+    let deduped_ids = dedup_ids(ids);
+
+    provider.subscribe(&deduped_ids).await?;
+    query_websocket(&mut provider, deduped_ids, timeout).await
+}
+
+async fn query_websocket_in_chunks<C, P, E1, E2>(
+    connector: C,
+    ids: &[String],
+    max_subscription_per_connection: usize,
+    timeout: Duration,
+) -> anyhow::Result<Vec<AssetInfo>>
+where
+    E1: StdError + Send + Sync + 'static,
+    E2: StdError + Send + Sync + 'static,
+    P: WebSocketAssetInfoProvider<SubscriptionError = E1, PollingError = E2>,
+    C: AssetInfoProviderConnector<Provider = P, Error = E1>,
+{
+    let mut tasks = FuturesOrdered::new();
+    let deduped_ids = dedup_ids(ids);
+
+    for chunk in deduped_ids
+        .into_iter()
+        .chunks(max_subscription_per_connection)
+        .into_iter()
+    {
+        let chunk_ids = chunk.collect::<Vec<String>>();
+        let mut provider = connector.connect().await?;
+        WebSocketAssetInfoProvider::subscribe(&mut provider, &chunk_ids).await?;
+
+        tasks.push_back(async move { query_websocket(&mut provider, chunk_ids, timeout).await });
+    }
+
+    let asset_infos = tasks
+        .try_collect::<Vec<Vec<AssetInfo>>>()
+        .await?
+        .into_iter()
+        .flatten()
+        .collect();
+
+    Ok(asset_infos)
+}
+
+async fn query_websocket<E1, E2, P>(
+    provider: &mut P,
+    ids: Vec<String>,
```

Comment:
if all we're doing is matching the hashmap size here ids is unnecessary since subscribe is done outside this scope.